src/phoenix/experimental/evals/functions/executor.py

```diff
@@ -0,0 +1,338 @@
+        generates tasks to be executed.
+
+        concurrency (int, optional): The number of concurrent consumers. Defaults to 3.
+
+        tqdm_bar_format (Optional[str], optional): The format string for the progress bar. Defaults
+        to None.
+
+        max_retries (int, optional): The maximum number of times to retry on exceptions. Defaults to
+        10.
+
+        exit_on_error (bool, optional): Whether to exit execution on the first encountered error.
+        Defaults to True.
+
+        fallback_return_value (Union[Unset, Any], optional): The fallback return value for tasks
+        that encounter errors. Defaults to _unset.
+    """
+
+    def __init__(
+        self,
+        generation_fn: Callable[[Any], Coroutine[Any, Any, Any]],
+        concurrency: int = 3,
+        tqdm_bar_format: Optional[str] = None,
+        max_retries: int = 10,
+        exit_on_error: bool = True,
+        fallback_return_value: Union[Unset, Any] = _unset,
+    ):
+        self.generate = generation_fn
+        self.fallback_return_value = fallback_return_value
+        self.concurrency = concurrency
+        self.tqdm_bar_format = tqdm_bar_format
+        self.max_retries = max_retries
+        self.exit_on_error = exit_on_error
+        self.base_priority = 0
+
+    async def producer(
+        self,
+        inputs: Sequence[Any],
+        queue: asyncio.PriorityQueue[Tuple[int, Any]],
+        max_fill: int,
+        done_producing: asyncio.Event,
+        termination_signal: asyncio.Event,
+    ) -> None:
+        try:
+            for index, input in enumerate(inputs):
+                if termination_signal.is_set():
+                    break
+                while queue.qsize() >= max_fill:
+                    # keep room in the queue for requeues
+                    await asyncio.sleep(1)
+                await queue.put((self.base_priority, (index, input)))
+        finally:
+            done_producing.set()
+
+    async def consumer(
+        self,
+        output: List[Any],
+        queue: asyncio.PriorityQueue[Tuple[int, Any]],
+        done_producing: asyncio.Event,
+        termination_signal: asyncio.Event,
+        progress_bar: tqdm[Any],
+    ) -> None:
+        termination_signal_task = None
+        while True:
+            marked_done = False
+            try:
+                priority, item = await asyncio.wait_for(queue.get(), timeout=1)
+            except asyncio.TimeoutError:
+                if done_producing.is_set() and queue.empty():
+                    break
+                continue
+            if termination_signal.is_set():
+                # discard any remaining items in the queue
+                queue.task_done()
+                marked_done = True
+                continue
+
+            index, payload = item
+            try:
+                generate_task = asyncio.create_task(self.generate(payload))
+                termination_signal_task = asyncio.create_task(termination_signal.wait())
```

Comment:
Termination events should not be attached to the `Executor` class, as it will be instantiated outside of the event loop created by its own `AsyncExecutor.run` method. By moving the creation of this event inside we ensure that the `Event` is created in the correct loop context.